{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c81098c8",
   "metadata": {
    "notebookRunGroups": {
     "groupValue": "1"
    }
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "# 转换plink格式\n",
    "plink --vcf input.vcf \\\n",
    "    --autosome-num 42 \\\n",
    "    --recode \\\n",
    "    --out output \\\n",
    "    --double-id\n",
    "\n",
    "# plink ped to bed格式\n",
    "plink --file FILENAME \\\n",
    "    --make-bed \\\n",
    "    --chr-set 42 \\\n",
    "    --out FILENAME\n",
    "\n",
    "# Plink PCA\n",
    "plink --bfile INPUT \\\n",
    "    --pca 10 \\\n",
    "    --out OUTPUT\n",
    "\n",
    "# 查看缺失率\n",
    "plink --file input \\\n",
    "    --autosome-num 42 \\\n",
    "    -missing\n",
    "\n",
    "# 查看maf\n",
    "plink --file output \\\n",
    "    --autosome-num 42 \\\n",
    "    --freq \\\n",
    "    --out MAF_check\n",
    "\n",
    "# 查看杂合度\n",
    "plink --file output \\\n",
    "    --het \\\n",
    "    --out R_check"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f15896e2",
   "metadata": {
    "tags": [
     "r"
    ]
   },
   "outputs": [],
   "source": [
    "%%r\n",
    "\n",
    "# R画缺失率\n",
    "indmiss<-read.table(file=\"plink.imiss\", header=TRUE)\n",
    "snpmiss<-read.table(file=\"plink.lmiss\", header=TRUE)\n",
    "pdf(\"imiss.pdf\")\n",
    "hist(indmiss[,6],main=\"Histogram individual missingness\")\n",
    "pdf(\"lmiss.pdf\")\n",
    "hist(snpmiss[,5],main=\"Histogram SNP missingness\")\n",
    "dev.off()\n",
    "\n",
    "# R画maf图\n",
    "maf_freq <- read.table(\"MAF_check.frq\", header =TRUE, as.is=T)\n",
    "pdf(\"MAF_distribution.pdf\")\n",
    "hist(maf_freq[,5],main = \"MAF distribution\", xlab = \"MAF\")\n",
    "dev.off()\n",
    "\n",
    "# R画杂合度\n",
    "het <- read.table(\"heter_check.het\", head=TRUE)\n",
    "pdf(\"heterozygosity.pdf\")\n",
    "het$HET_RATE = (het$\"N.NM.\" - het$\"O.HOM.\")/het$\"N.NM.\"\n",
    "hist(het$HET_RATE, xlab=\"Heterozygosity Rate\", ylab=\"Frequency\", main= \"Heterozygosity Rate\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac821f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "vcftools --vcf chr${i}_noID.vcf \\\n",
    "    --max-missing 0.95 \\\n",
    "    --remove-indels \\\n",
    "    --maf 0.05 \\\n",
    "    --min-alleles 2 \\\n",
    "    --max-alleles 2 \\\n",
    "    --mac 2 \\\n",
    "    --recode \\\n",
    "    --recode-INFO-all \\\n",
    "    --out chr${i}_filter\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
