{
 "cells": [
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# 1 Tried to run\n",
    "\n",
    "## 1.1 samtools version error\n",
    "\n",
    "The `FastCall2` step `disc` will report an error if the `samtools` version is not compatible. There are 3 widely used versions of `samtools`:\n",
    "\n",
    "1. `Version: 1.18 (using htslib 1.21)`\n",
    "```\n",
    "Input options:\n",
    "  --rf, --incl-flags STR|INT\n",
    "                          required flags: only include reads with any of\n",
    "                          the mask bits set []\n",
    "  --ff, --excl-flags STR|INT\n",
    "                          filter flags: skip reads with any of the mask bits set\n",
    "                                            [UNMAP,SECONDARY,QCFAIL,DUP]\n",
    "  -x, --ignore-overlaps-removal, --disable-overlap-removal\n",
    "                          disable read-pair overlap detection and removal\n",
    "  -X, --customized-index  use customized index files\n",
    "\n",
    "Output options:\n",
    "  -O, --output-BP          output base positions on reads, current orientation\n",
    "      --output-BP-5        output base positions on reads, 5' to 3' orientation\n",
    "  -M, --output-mods        output base modifications\n",
    "  -s, --output-MQ          output mapping quality\n",
    "      --output-QNAME       output read names\n",
    "      --output-extra STR   output extra read fields and read tag values\n",
    "      --output-sep CHAR    set the separator character for tag lists [,]\n",
    "      --output-empty CHAR  set the no value character for tag lists [*]\n",
    "      --no-output-ins      skip insertion sequence after +NUM\n",
    "                           Use twice for complete insertion removal\n",
    "      --no-output-ins-mods don't display base modifications within insertions\n",
    "      --no-output-del      skip deletion sequence after -NUM\n",
    "                           Use twice for complete deletion removal\n",
    "      --no-output-ends     remove ^MQUAL and $ markup in sequence column\n",
    "      --reverse-del        use '#' character for deletions on the reverse strand\n",
    "\n",
    "Generic options:\n",
    "      --verbosity INT\n",
    "               Set level of verbosity\n",
    "\n",
    "Note that using \"samtools mpileup\" to generate BCF or VCF files has been\n",
    "removed.  To output these formats, please use \"bcftools mpileup\" instead.\n",
    "```\n",
    "2. `Version: 1.9 (using htslib 1.9)`\n",
    "```\n",
    "Input options:\n",
    "  --rf, --incl-flags STR|INT  required flags: skip reads with mask bits unset []\n",
    "  --ff, --excl-flags STR|INT  filter flags: skip reads with mask bits set\n",
    "                                            [UNMAP,SECONDARY,QCFAIL,DUP]\n",
    "  -x, --ignore-overlaps   disable read-pair overlap detection\n",
    "\n",
    "Output options:\n",
    "  -O, --output-BP         output base positions on reads\n",
    "  -s, --output-MQ         output mapping quality\n",
    "      --output-QNAME      output read names\n",
    "\n",
    "Generic options:\n",
    "\n",
    "Note that using \"samtools mpileup\" to generate BCF or VCF files is now\n",
    "deprecated.  To output these formats, please use \"bcftools mpileup\" instead.\n",
    "```\n",
    "3. `Version: 1.21 (using htslib 1.21)`\n",
    "```\n",
    "Input options:\n",
    "  --rf, --incl-flags STR|INT\n",
    "                          required flags: only include reads with any of\n",
    "                          the mask bits set []\n",
    "  --ff, --excl-flags STR|INT\n",
    "                          filter flags: skip reads with any of the mask bits set\n",
    "                                            [UNMAP,SECONDARY,QCFAIL,DUP]\n",
    "  -x, --ignore-overlaps-removal, --disable-overlap-removal\n",
    "                          disable read-pair overlap detection and removal\n",
    "  -X, --customized-index  use customized index files\n",
    "\n",
    "Output options:\n",
    "  -O, --output-BP          output base positions on reads, current orientation\n",
    "      --output-BP-5        output base positions on reads, 5' to 3' orientation\n",
    "  -M, --output-mods        output base modifications\n",
    "  -s, --output-MQ          output mapping quality\n",
    "      --output-QNAME       output read names\n",
    "      --output-extra STR   output extra read fields and read tag values\n",
    "      --output-sep CHAR    set the separator character for tag lists [,]\n",
    "      --output-empty CHAR  set the no value character for tag lists [*]\n",
    "      --no-output-ins      skip insertion sequence after +NUM\n",
    "                           Use twice for complete insertion removal\n",
    "      --no-output-ins-mods don't display base modifications within insertions\n",
    "      --no-output-del      skip deletion sequence after -NUM\n",
    "                           Use twice for complete deletion removal\n",
    "      --no-output-ends     remove ^MQUAL and $ markup in sequence column\n",
    "      --reverse-del        use '#' character for deletions on the reverse strand\n",
    "\n",
    "Generic options:\n",
    "      --verbosity INT\n",
    "               Set level of verbosity\n",
    "\n",
    "Note that using \"samtools mpileup\" to generate BCF or VCF files has been\n",
    "removed.  To output these formats, please use \"bcftools mpileup\" instead.\n",
    "```\n",
    "The command line version of FastCall2 using `samtools` is: `samtools mpileup --no-output-ends -B -q 30 -Q 20 -f chr001.fa taxon.bam -r 1`. Therefore, the error is caused by the `--no-output-ends` option, which is not supported in `samtools 1.9`.\n",
    "\n",
    "1.21快20%左右\n",
    "\n",
    "The existing `samtools` version includes:\n",
    "```\n",
    "# Name                       Version           Build  Channel\n",
    "samtools                      0.1.12               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.12               1  anaconda/cloud/bioconda\n",
    "samtools                      0.1.12               2  anaconda/cloud/bioconda\n",
    "samtools                      0.1.13               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.13               1  anaconda/cloud/bioconda\n",
    "samtools                      0.1.14               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.14               1  anaconda/cloud/bioconda\n",
    "samtools                      0.1.15               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.16               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.17               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18     h20b1175_12  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18     h270b39a_11  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18     h50ea8bc_13  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18     h96c455f_14  anaconda/cloud/bioconda\n",
    "samtools                      0.1.18     hfb9b9cc_10  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19               0  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19               1  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19               2  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19               3  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19     h20b1175_10  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19      h270b39a_9  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19     h50ea8bc_11  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19      h94a8ba4_5  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19      h94a8ba4_6  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19     h96c455f_12  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19     h96c455f_13  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19      hf89b575_7  anaconda/cloud/bioconda\n",
    "samtools                      0.1.19      hfb9b9cc_8  anaconda/cloud/bioconda\n",
    "samtools                         1.0               0  anaconda/cloud/bioconda\n",
    "samtools                         1.0               1  anaconda/cloud/bioconda\n",
    "samtools                         1.0      hdd8ed8b_2  anaconda/cloud/bioconda\n",
    "samtools                         1.1               0  anaconda/cloud/bioconda\n",
    "samtools                         1.1      h20b1175_7  anaconda/cloud/bioconda\n",
    "samtools                         1.1      h270b39a_6  anaconda/cloud/bioconda\n",
    "samtools                         1.1      h50ea8bc_8  anaconda/cloud/bioconda\n",
    "samtools                         1.1      h94a8ba4_3  anaconda/cloud/bioconda\n",
    "samtools                         1.1      h96c455f_9  anaconda/cloud/bioconda\n",
    "samtools                         1.1      hf89b575_4  anaconda/cloud/bioconda\n",
    "samtools                         1.1      hfb9b9cc_5  anaconda/cloud/bioconda\n",
    "samtools                   1.2.rglab               0  anaconda/cloud/bioconda\n",
    "samtools                         1.2               0  anaconda/cloud/bioconda\n",
    "samtools                         1.2               1  anaconda/cloud/bioconda\n",
    "samtools                         1.2               2  anaconda/cloud/bioconda\n",
    "samtools                         1.3               0  anaconda/cloud/bioconda\n",
    "samtools                         1.3               1  anaconda/cloud/bioconda\n",
    "samtools                         1.3               2  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h0592bc0_3  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h0cf4675_8  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h1b8c3c0_5  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h41a57b0_6  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h60f3df9_9  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h80b0bb3_4  anaconda/cloud/bioconda\n",
    "samtools                         1.3      h9071d68_7  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               0  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               1  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               2  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               3  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               4  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1               5  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1      h0592bc0_6  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1     h0cf4675_11  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1      h1b8c3c0_8  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1      h41a57b0_9  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1     h60f3df9_12  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1      h80b0bb3_7  anaconda/cloud/bioconda\n",
    "samtools                       1.3.1     h9071d68_10  anaconda/cloud/bioconda\n",
    "samtools                         1.4               0  anaconda/cloud/bioconda\n",
    "samtools                       1.4.1               0  anaconda/cloud/bioconda\n",
    "samtools                         1.5               0  anaconda/cloud/bioconda\n",
    "samtools                         1.5               1  anaconda/cloud/bioconda\n",
    "samtools                         1.5               2  anaconda/cloud/bioconda\n",
    "samtools                         1.6               0  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h02bfda8_2  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h244ad75_4  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h244ad75_5  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h3f2fef4_8  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h4da6232_1  anaconda/cloud/bioconda\n",
    "samtools                         1.6     h5fe306e_11  anaconda/cloud/bioconda\n",
    "samtools                         1.6     h5fe306e_12  anaconda/cloud/bioconda\n",
    "samtools                         1.6      h9dace67_6  anaconda/cloud/bioconda\n",
    "samtools                         1.6      hb116620_7  anaconda/cloud/bioconda\n",
    "samtools                         1.6     hc3601fc_10  anaconda/cloud/bioconda\n",
    "samtools                         1.6      hcd7b337_9  anaconda/cloud/bioconda\n",
    "samtools                         1.6      he673b24_3  anaconda/cloud/bioconda\n",
    "samtools                         1.7               0  anaconda/cloud/bioconda\n",
    "samtools                         1.7               1  anaconda/cloud/bioconda\n",
    "samtools                         1.7               2  anaconda/cloud/bioconda\n",
    "samtools                         1.8               2  anaconda/cloud/bioconda\n",
    "samtools                         1.8               3  anaconda/cloud/bioconda\n",
    "samtools                         1.8               4  anaconda/cloud/bioconda\n",
    "samtools                         1.8      h46bd0b3_5  anaconda/cloud/bioconda\n",
    "samtools                         1.9     h10a08f8_12  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h43f6869_9  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h46bd0b3_0  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h57cc563_6  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h57cc563_7  anaconda/cloud/bioconda\n",
    "samtools                         1.9     h8571acd_10  anaconda/cloud/bioconda\n",
    "samtools                         1.9     h8571acd_11  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h8ee4bcc_1  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h91753b0_2  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h91753b0_3  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h91753b0_4  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h91753b0_5  anaconda/cloud/bioconda\n",
    "samtools                         1.9      h91753b0_8  anaconda/cloud/bioconda\n",
    "samtools                        1.10      h2e538c0_3  anaconda/cloud/bioconda\n",
    "samtools                        1.10      h9402c20_0  anaconda/cloud/bioconda\n",
    "samtools                        1.10      h9402c20_1  anaconda/cloud/bioconda\n",
    "samtools                        1.10      h9402c20_2  anaconda/cloud/bioconda\n",
    "samtools                        1.11      h6270b1f_0  anaconda/cloud/bioconda\n",
    "samtools                        1.12      h9aed4be_1  anaconda/cloud/bioconda\n",
    "samtools                        1.12      hd5e65b6_0  anaconda/cloud/bioconda\n",
    "samtools                        1.13      h8c37831_0  anaconda/cloud/bioconda\n",
    "samtools                        1.14      hb421002_0  anaconda/cloud/bioconda\n",
    "samtools                        1.15      h1170115_1  anaconda/cloud/bioconda\n",
    "samtools                        1.15      h3843a85_0  anaconda/cloud/bioconda\n",
    "samtools                      1.15.1      h1170115_0  anaconda/cloud/bioconda\n",
    "samtools                      1.15.1      h6899075_1  anaconda/cloud/bioconda\n",
    "samtools                      1.16.1      h00cdaf9_2  anaconda/cloud/bioconda\n",
    "samtools                      1.16.1      h6899075_0  anaconda/cloud/bioconda\n",
    "samtools                      1.16.1      h6899075_1  anaconda/cloud/bioconda\n",
    "samtools                        1.17      h00cdaf9_0  anaconda/cloud/bioconda\n",
    "samtools                        1.17      hd87286a_1  anaconda/cloud/bioconda\n",
    "samtools                        1.17      hd87286a_2  anaconda/cloud/bioconda\n",
    "samtools                        1.18      h50ea8bc_1  anaconda/cloud/bioconda\n",
    "samtools                        1.18      hd87286a_0  anaconda/cloud/bioconda\n",
    "samtools                        1.19      h50ea8bc_0  anaconda/cloud/bioconda\n",
    "samtools                      1.19.1      h50ea8bc_0  anaconda/cloud/bioconda\n",
    "samtools                      1.19.2      h50ea8bc_0  anaconda/cloud/bioconda\n",
    "samtools                      1.19.2      h50ea8bc_1  anaconda/cloud/bioconda\n",
    "samtools                        1.20      h50ea8bc_0  anaconda/cloud/bioconda\n",
    "samtools                        1.20      h50ea8bc_1  anaconda/cloud/bioconda\n",
    "samtools                        1.21      h50ea8bc_0  anaconda/cloud/bioconda\n",
    "samtools                        1.21      h96c455f_1  anaconda/cloud/bioconda\n",
    "samtools                        1.22      h96c455f_0  anaconda/cloud/bioconda\n",
    "```\n"
   ],
   "id": "91aa054d871a1cb7"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## 1.2 scan step error\n",
    "\n",
    "Run `scan` with command line: `java -Xmx100g -jar TIGER_20250526.jar -app FastCall2 -mod scan -a ../00data/chr1_10M.fa -b taxaBamMap.txt -c ../blib/1_1_10000001.lib.gz -d 1 -e 0 -f 30 -g 20 -h 0.05 -i /data/home/dazheng/miniconda3/envs/debug/bin/samtools -j 32 -k ../scan`\n",
    "\n",
    "I got the following error:\n",
    "```\n",
    "java.lang.ArrayIndexOutOfBoundsException: 292652\n",
    "        at pgl.app.fastCall2.ScanGenotype.lambda$mkFinalVCFFromIndiCounts$0(ScanGenotype.java:274)\n",
    "        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:183)\n",
    "        at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1382)\n",
    "        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)\n",
    "        at java.util.stream.ForEachOps$ForEachTask.compute(ForEachOps.java:290)\n",
    "        at java.util.concurrent.CountedCompleter.exec(CountedCompleter.java:731)\n",
    "        at java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)\n",
    "        at java.util.concurrent.ForkJoinTask.doInvoke(ForkJoinTask.java:401)\n",
    "        at java.util.concurrent.ForkJoinTask.invoke(ForkJoinTask.java:734)\n",
    "        at java.util.stream.ForEachOps$ForEachOp.evaluateParallel(ForEachOps.java:159)\n",
    "        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateParallel(ForEachOps.java:173)\n",
    "        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:233)\n",
    "        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)\n",
    "        at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:650)\n",
    "        at pgl.app.fastCall2.ScanGenotype.mkFinalVCFFromIndiCounts(ScanGenotype.java:262)\n",
    "        at pgl.app.fastCall2.ScanGenotype.<init>(ScanGenotype.java:89)\n",
    "        at pgl.app.fastCall2.FastCall2.<init>(FastCall2.java:55)\n",
    "        at pgl.PGLAPPEntrance.runApp(PGLAPPEntrance.java:51)\n",
    "        at pgl.PGLAPPEntrance.<init>(PGLAPPEntrance.java:27)\n",
    "        at pgl.PGLAPPEntrance.main(PGLAPPEntrance.java:85)\n",
    "```\n",
    "\n",
    "When I tried to run `disc` for 4806 samples, I got the error message:\n",
    "```\n",
    "java.io.FileNotFoundException: /data/home/tusr1/02callSNP/lib/../disc/Becard/Kachu-2/1_1_5000001.ing.gz (No such file or directory)\n",
    "        at java.io.FileOutputStream.open0(Native Method)\n",
    "        at java.io.FileOutputStream.open(FileOutputStream.java:270)\n",
    "        at java.io.FileOutputStream.<init>(FileOutputStream.java:213)\n",
    "        at java.io.FileOutputStream.<init>(FileOutputStream.java:101)\n",
    "        at pgl.infra.utils.IOUtils.getBinaryGzipWriter(IOUtils.java:115)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.setDos(DiscoverVariation.java:384)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.writeVariants(DiscoverVariation.java:400)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.call(DiscoverVariation.java:422)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.call(DiscoverVariation.java:243)\n",
    "        at java.util.concurrent.FutureTask.run(FutureTask.java:266)\n",
    "        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
    "        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
    "        at java.lang.Thread.run(Thread.java:750)\n",
    "java.lang.NullPointerException\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.setDos(DiscoverVariation.java:386)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.writeVariants(DiscoverVariation.java:400)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.call(DiscoverVariation.java:422)\n",
    "        at pgl.app.fastCall2.DiscoverVariation$TaxonCall.call(DiscoverVariation.java:243)\n",
    "        at java.util.concurrent.FutureTask.run(FutureTask.java:266)\n",
    "        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)\n",
    "        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)\n",
    "        at java.lang.Thread.run(Thread.java:750)\n",
    "```\n",
    "This is because `@SM` of some samples contains `/`, which is not allowed in the file name. Therefore, I finally choose `CRR` code as the `Taxa` in `taxaBamMap.txt` file."
   ],
   "id": "7aa519d730581b71"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "## 1.3 multiple chromosomes in one shell script error\n",
    "\n",
    "When I tried to run `FastCall2` for multiple chromosomes in one shell script, I got the log information as follows:\n",
    "1. The first chromosome `3` runs successfully, with\n",
    "    - `Reading reference genome took 982.73s`\n",
    "    - `FastCall 2 is finished in 12.310276 hours.`\n",
    "2. But other chromosomes, such as `4`, `5`, `6`, and `7`, report the error as the program directly end after reading reference genome: \n",
    "    - `Reading reference genome took 1029.84s`\n",
    "    - `FastCall 2 is finished in 0.28940982 hours.`\n",
    "\n",
    "没有输出报错文件，捕获到\n",
    "\n",
    "测试一下\n",
    "\n",
    "To find the reason for this, I build a test dataset for wap population with 2012 samples. Each bam file of the test dataset contains reads for each chromosome at the first 1M region.\n",
    "\n",
    "\n"
   ],
   "id": "80f60fff8be5a4a4"
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": [
    "# 2 perf analysis\n",
    "\n"
   ],
   "id": "97d9eefe9039c843"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
